render/pixman/renderer.c

```diff
@@ -0,0 +1,410 @@
+#include "render/pixman.h"
+
+#include <assert.h>
+#include <drm_fourcc.h>
+#include <pixman.h>
+#include <stdlib.h>
+#include <wayland-server.h>
+#include <wlr/render/interface.h>
+#include <wlr/types/wlr_matrix.h>
+#include <wlr/util/log.h>
+
+#include "types/wlr_buffer.h"
+
+static const struct wlr_renderer_impl renderer_impl;
+
+static struct wlr_pixman_renderer *get_renderer(
+		struct wlr_renderer *wlr_renderer) {
+	assert(wlr_renderer->impl == &renderer_impl);
+	return (struct wlr_pixman_renderer *)wlr_renderer;
+}
+
+static struct wlr_pixman_buffer *get_buffer(
+		struct wlr_pixman_renderer *renderer, struct wlr_buffer *wlr_buffer) {
+	struct wlr_pixman_buffer *buffer;
+	wl_list_for_each(buffer, &renderer->buffers, link) {
+		if (buffer->buffer == wlr_buffer) {
+			return buffer;
+		}
+	}
+	return NULL;
+}
+
+static const struct wlr_texture_impl texture_impl;
+
+static struct wlr_pixman_texture *get_texture(
+		struct wlr_texture *wlr_texture) {
+	assert(wlr_texture->impl == &texture_impl);
+	return (struct wlr_pixman_texture *)wlr_texture;
+}
+
+static bool texture_is_opaque(struct wlr_texture *wlr_texture) {
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+	wlr_log(WLR_INFO, "texture_is_opaque");
+
+#if WLR_BIG_ENDIAN
+	return texture->format == PIXMAN_b8g8r8x8;
+#else
+	return texture->format == PIXMAN_x8r8g8b8;
+#endif
+}
+
+static bool texture_write_pixels(struct wlr_texture *wlr_texture,
+		uint32_t stride, uint32_t width, uint32_t height,
+		uint32_t src_x, uint32_t src_y, uint32_t dst_x, uint32_t dst_y,
+		const void *data) {
+	wlr_log(WLR_INFO, "texture_write_pixels");
+
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+
+	wlr_log(WLR_INFO, "tex is w %zu h %zu", (size_t)wlr_texture->width,
+			(size_t)wlr_texture->height);
+
+	wlr_log(WLR_INFO, "stride is %zu", (size_t)stride);
+
+	wlr_log(WLR_INFO, "src_x %zu src_y %zu", (size_t)src_x, (size_t)src_y);
+
+	uint32_t *start = (uint32_t*)data + (src_y * stride) + (src_x * 4);
+	assert(start);
+
+	pixman_image_t *pixels = pixman_image_create_bits_no_clear(texture->format,
+			width, height, start, stride);
+
+	pixman_image_composite32(PIXMAN_OP_OVER, texture->image, NULL, pixels,
+			dst_x, dst_y, 0, 0, 0, 0, width, height);
```

Comment:
I don't think we should do this: `texture->image` points to client memory, we shouldn't write to client buffers.

We can probably just `return false` and the wlr_buffer code will re-create the texture (which isn't particularly expensive since no buffer copying is involved).

Reply:
Even better: we can just leave `write_pixels` unimplemented.